tortoise/verifying_tortoise.go

```diff
@@ -809,27 +809,164 @@ func (t *turtle) handleLayerBlocks(ctx *tcontext, layerID types.LayerID) error {
 	}
 	return t.processBlocks(ctx, layerBlocks)
 }
 
+func (t *turtle) verifyingTortoise(ctx *tcontext, logger log.Log, lid types.LayerID) (map[types.BlockID]bool, error) {
+	localOpinion, err := t.getLocalOpinion(ctx, lid)
+	if err != nil {
+		return nil, err
+	}
+	if len(localOpinion) == 0 {
+		// warn about this to be safe, but we do allow empty layers and must be able to verify them
+		logger.With().Warning("empty vote vector for layer", lid)
+	}
+
+	contextualValidity := make(map[types.BlockID]bool, len(localOpinion))
+
+	// Count the votes of good blocks. localOpinionOnBlock is our local opinion on this block.
+	// Declare the vote vector "verified" up to position k if the total weight exceeds the confidence threshold in
+	// all positions up to k: in other words, we can verify a layer k if the total weight of the global opinion
+	// exceeds the confidence threshold, and agrees with local opinion.
+	for bid, localOpinionOnBlock := range localOpinion {
+		// count the votes of the input vote vector by summing the voting weight of good blocks
+		logger.With().Debug("summing votes for candidate layer block",
+			bid,
+			log.FieldNamed("layer_start", lid.Add(1)),
+			log.FieldNamed("layer_end", t.Last),
+		)
+		sum, err := t.sumVotesForBlock(ctx, bid, lid.Add(1), func(votingBlockID types.BlockID) bool {
+			if _, isgood := t.GoodBlocksIndex[votingBlockID]; !isgood {
+				logger.With().Debug("not counting vote of block not marked good",
+					log.FieldNamed("voting_block", votingBlockID))
+				return false
+			}
+			return true
+		})
+		if err != nil {
+			return nil, fmt.Errorf("error summing votes for block %v in candidate layer %v: %w",
+				bid, lid, err)
+		}
+
+		// check that the total weight exceeds the global threshold
+		globalOpinionOnBlock := calculateOpinionWithThreshold(
+			t.logger, sum, t.GlobalThreshold, t.AvgLayerSize, t.Last.Difference(lid))
+		logger.With().Debug("verifying tortoise calculated global opinion on block",
+			log.FieldNamed("block_voted_on", bid),
+			lid,
+			log.FieldNamed("global_vote_sum", sum),
+			log.FieldNamed("global_opinion", globalOpinionOnBlock),
+			log.FieldNamed("local_opinion", localOpinionOnBlock))
+
+		// At this point, we have all of the data we need to make a decision on this block. There are three possible
+		// outcomes:
+		// 1. record our opinion on this block and go on evaluating the rest of the blocks in this layer to see if
+		//    we can verify the layer (if local and global consensus match, and global consensus is decided)
+		// 2. keep waiting to verify the layer (if not, and the layer is relatively recent)
+		// 3. trigger self healing (if not, and the layer is sufficiently old)
+		consensusMatches := globalOpinionOnBlock == localOpinionOnBlock
+		globalOpinionDecided := globalOpinionOnBlock != abstain
+
+		if consensusMatches && globalOpinionDecided {
+			// Opinion on this block is decided, save and keep going
+			contextualValidity[bid] = globalOpinionOnBlock == support
```

Comment:
It looks to me that we can move these two lines to the end of the function before `return`, which would let us remove the `if` statement above. Is that correct?